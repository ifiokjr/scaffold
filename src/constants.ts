/**
 * The version of the CLI.
 */
export const VERSION = /* ={version} */ "0.3.0" /* {/version} */;

export const SUPPORTED_EXTENSIONS = [
  ".ts",
  ".js",
  ".mts",
  ".mjs",
  ".tsx",
  ".jsx",
] as const;
