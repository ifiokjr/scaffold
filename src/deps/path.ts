export { default as normalizePath } from "https://cdn.skypack.dev/normalize-path@v3.0.0?dts";
export * as path from "https://deno.land/std@0.144.0/path/mod.ts";
export * from "https://esm.sh/junk@4.0.0?type=deno";
