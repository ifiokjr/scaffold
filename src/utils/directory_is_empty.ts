/**
 * Check whether the directory is empty while ignoring whether the directory exists.
 */
export async function directoryIsEmpty(
  directory: string | URL,
): Promise<boolean> {
  const path = typeof directory === "string" ? directory : directory.pathname;

  try {
    for await (const _ of Deno.readDir(path)) {
      return false;
    }
  } catch {
    return true;
  }

  return true;
}
