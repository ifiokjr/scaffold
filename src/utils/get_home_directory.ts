const [windowsStatus, unixStatus] = await Promise.all([
  Deno.permissions.query({ name: "env", variable: "USERPROFILE" }),
  Deno.permissions.query({ name: "env", variable: "HOME" }),
]);

const canCheckWindows = windowsStatus.state === "granted";
const canCheckUnix = unixStatus.state === "granted";

/** Returns the string path of the current user's home directory. Taken from
 * https://github.com/denoland/deno_std/blob/5c26306183d0a57f4ae1f4264201b7956526aa01/node/os.ts#L160-L174
 *
 * Requires the `--allow-env "USERPROFILE"` and `--allow-env "HOME"` options to
 * be passed to Deno.
 */
export function getHomeDirectory(): string | undefined {
  if (Deno.build.os === "windows" && canCheckWindows) {
    return Deno.env.get("USERPROFILE");
  }

  if (canCheckUnix) {
    return Deno.env.get("HOME");
  }

  return;
}
