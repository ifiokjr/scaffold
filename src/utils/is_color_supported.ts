let supported = true;

const permissions = await Promise.all([
  Deno.permissions.query({ name: "env", variable: "TERM" }),
  Deno.permissions.query({ name: "env", variable: "CI" }),
  Deno.permissions.query({ name: "env", variable: "FORCE_COLOR" }),
]);

if (permissions.every((permission) => permission.state === "granted")) {
  supported =
    // terminal is not dumb
    (Deno.isatty(1) && Deno.env.get("TERM") !== "dumb") ||
    // default support in CI systems which haven't disabled colors
    !!Deno.env.get("CI") ||
    // colors are forced by the environment
    !!Deno.env.get("FORCE_COLOR");
}

/**
 * Check for color support.
 *
 * Requires the `--allow-env` flag to be passed to Deno.
 */
export function isColorSupported() {
  // Exit early if the user has disabled colors.
  if (Deno.noColor || Deno.args.includes("--no-color")) {
    return false;
  }

  return (
    // forced by args used when calling this script
    Deno.args.includes("--color") ||
    // support colors on windows by default
    Deno.build.os === "windows" || supported
  );
}
