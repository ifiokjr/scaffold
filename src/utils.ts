import { colors, Spinner, type SpinnerOptions } from "./deps/cli.ts";
import { normalizePath, path } from "./deps/path.ts";

/**
 * Removes all undefined values from an object. Neither Firestore nor the
 * RealtimeDB allow `undefined` as a value.
 *
 * @param data The object to clean
 * @param nullish Set to true to clean null and undefined values
 */
export function removeUndefined<Shape extends object>(
  data: Shape,
  nullish = false,
) {
  const transformed = Object.create({});

  for (const [key, value] of Object.entries(data)) {
    if (nullish ? value == null : value === undefined) {
      continue;
    }

    transformed[key] = value;
  }

  return transformed;
}

const encoder = new TextEncoder();

/**
 * Encode a string for passing through to an anchor program.
 */
export function stringToUint8Array(text: string): Uint8Array {
  return encoder.encode(text);
}

const decoder = new TextDecoder();

/**
 * Decode a Uint8Array to a string.
 */
export function uint8ArrayToString(input: Uint8Array): string {
  return decoder.decode(input);
}

/**
 * Normalize a path to it's posix form.
 *
 * @param filePath The path to normalize
 * @param [stripTrailing=false] When set to true any trailing slashes will be
 * removed.
 *
 * @returns a normalized path with posix style separators
 */
export function normalize(filePath: string, stripTrailing?: boolean): string {
  return path.normalize(normalizePath(filePath, stripTrailing ?? false));
}

/**
 * Normalize a path as a directory with a trailing slash.
 *
 * @param directory The directory to normalize
 * @param [trailingSlash=true] When set to true a trailing slash will be added.
 */
export function normalizeDirectory(directory: string, trailingSlash = true) {
  return `${normalize(directory, true)}${trailingSlash ? "/" : ""}`;
}

/**
 * Get the normalized path of the provided `path`.
 */
export function getPath(filepath: string | URL): string {
  return filepath instanceof URL
    ? filepath.pathname
    : filepath.startsWith("file:")
    ? new URL(filepath).pathname
    : normalize(filepath);
}

interface ReadJsonOptions {
  /**
   * When set to true, the file with the default data will be created if it does
   * not exist.
   */
  create?: boolean;

  /**
   * The default data to use when the file does not exist.
   *
   * @default {}
   */
  defaultData?: object;
}

export async function readJson(path: string, options: ReadJsonOptions = {}) {
  const { create = false, defaultData = Object.create(null) } = options;

  try {
    return JSON.parse(await Deno.readTextFile(path));
  } catch {
    if (create) {
      await Deno.writeTextFile(path, JSON.stringify(defaultData));
    }

    return defaultData;
  }
}

export async function writeJson(path: string, json: object) {
  try {
    const content = JSON.stringify(json);
    await Deno.writeTextFile(path, content, { create: true });
  } catch {
    return {};
  }
}

export function wait(options: SpinnerOptions) {
  let spinner: Spinner;
  const getSpinner = () => {
    return spinner ??= new Spinner({
      text: options.text,
      prefix: options.prefix ?? "",
      color: options.color ?? colors.cyan,
      spinner: options.spinner ?? "dots",
      hideCursor: options.hideCursor ?? true,
      indent: options.indent ?? 0,
      interval: options.interval ?? 100,
      stream: options.stream ?? Deno.stdout,
      enabled: true,
      discardStdin: true,
    }).start();
  };

  const proxy = {
    text: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().text = text;
      return proxy;
    },
    info: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().info(text);
      return proxy;
    },
    fail: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().fail(text);
      return proxy;
    },
    succeed: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().succeed(text);
      return proxy;
    },
    warn: (text: string) => {
      if (options.enabled === false) return;
      getSpinner().warn(text);
      return proxy;
    },
    stop: () => {
      if (options.enabled === false) return;
      getSpinner().stop();
      return proxy;
    },
  };

  return proxy;
}
