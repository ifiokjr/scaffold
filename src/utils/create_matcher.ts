import { isFunction, isString } from "../deps/npm.ts";
import { isJunk, path } from "../deps/path.ts";
import { globToRegExp, isGlob } from "../deps/std.ts";

export type MatchFunction = (source: string) => boolean;
export type Match = string | MatchFunction | RegExp;

export interface CreateMatcherOptions {
  /**
   * When a glob begins with `!` and the glob matches the source, the match will
   * be negated. Set this to `true` to disable the behavior.
   *
   * @default false
   */
  disableNegation?: boolean;

  /**
   * Expand the glob when it ends with `/`.
   *
   * @default false
   */
  expandGlobs?: boolean;

  /**
   * Set to true to make globs case insensitive.
   *
   * @default false
   */
  caseInsensitive?: boolean;

  /**
   * When true will match dot files and folders.
   *
   * @default false
   */
  dot?: boolean;

  /**
   * When true will match junk files
   *
   * @default false
   */
  junk?: boolean;

  /**
   * The extensions that can be matched. Setting this to anything other than an
   * empty array / or null will prevent matches on directories. Unless the
   * directory has an extension 🤷‍♀️.
   *
   * @default undefined
   */
  extensions?: string[] | null;
}
/**
 * Transform a pattern into a matching function.
 */

export function createMatcher(
  pattern: Match | Match[],
  options: CreateMatcherOptions = {},
): MatchFunction {
  const { disableNegation = false, expandGlobs = false } = options;
  const matchers = Array.isArray(pattern) ? pattern : [pattern].filter(Boolean);

  return (source) => {
    if (
      // Exit early if no matchers are provided.
      matchers.length === 0 ||
      // Exit early if the source is a dot file / folder.
      (
        !options.dot &&
        source.split(path.sep).some((part) => part.startsWith("."))
      ) ||
      // Exit early if the source is a junk file / folder.
      (
        !options.junk && (isJunk(source) || isJunk(path.basename(source)))
      ) ||
      // Exit early if not a valid extension.
      (
        options.extensions &&
        !options.extensions.some((extension) => source.endsWith(extension))
      )
    ) {
      return false;
    }

    let anyMatch = false;

    for (const matcher of matchers) {
      let match = false;
      // True when this is a string glob starting with `!`.
      let negated = false;

      if (isFunction(matcher)) {
        match = matcher(source) || match;
      } else if (isString(matcher) && !isGlob(matcher)) {
        match = matcher === source || match;
      } else {
        let regex: RegExp;

        if (isString(matcher)) {
          negated = disableNegation || matcher[0] === "!";
          regex = globToRegExp(
            `${negated ? matcher.slice(1) : matcher}${
              matcher.endsWith("/") && expandGlobs ? "**/*" : ""
            }`,
            { caseInsensitive: options.caseInsensitive },
          );
        } else {
          regex = matcher;
        }

        match = regex.test(source) || match;
      }

      if (match && disableNegation) {
        // exit early when a match is found and if negated return false.
        return true;
      }

      if (match && negated) {
        return false;
      }
      anyMatch = match || anyMatch;
    }

    return anyMatch;
  };
}
