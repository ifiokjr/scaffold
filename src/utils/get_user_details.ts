import { uint8ArrayToString } from "../utils.ts";

/**
 * Export the user details, where available.
 *
 * This requires `--allow-env="git"` and `--allow-env="whoami" to run.
 */
export async function getUserDetails() {
  const details = Object.create(null);
  const promises: Array<Promise<void>> = [];

  for (const [key, cmd] of Object.entries(commands)) {
    const promise = Deno.run({ cmd, stdout: "piped" }).output().then(
      (output) => {
        details[key] = uint8ArrayToString(output).trim() || undefined;
      },
    );

    promises.push(promise);
  }

  await Promise.all(promises);

  return details;
}

const commands: Record<keyof UserDetails, string[]> = {
  email: ["git", "config", "user.email"],
  username: ["git", "config", "user.name"],
  user: ["whoami"],
};

export interface UserDetails {
  /**
   * The email address of the user if available via `git`.
   */
  email: string | undefined;

  /**
   * The name via `git config user.name`.
   */
  username: string | undefined;

  /**
   * The name of the logged in user, if available use `whoami`
   */
  user: string | undefined;
}
