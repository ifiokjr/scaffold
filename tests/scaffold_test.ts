import { VERSION } from "../src/constants.ts";
import { path } from "../src/deps/path.ts";
import { uint8ArrayToString } from "../src/utils.ts";
import {
  afterAll,
  assertEquals,
  assertStringIncludes,
  describe,
  it,
} from "./deps.ts";
import { cwd, run, snapshotDirectory } from "./helpers.ts";

const needsRemoval: string[] = [];

afterAll(async () => {
  for (const file of needsRemoval) {
    try {
      await Deno.remove(file, { recursive: true });
    } catch {
      // ignore
    }
  }
});

describe("scaffold", () => {
  it("can display help", async () => {
    const command = run(["--help"], { stdout: "piped" });
    const output = uint8ArrayToString(await command.output());
    command.close();

    assertStringIncludes(output, "scaffold <repo> [folder]");
  });

  it("can display the version", async () => {
    const command = run(["-V"], { stdout: "piped" });
    const output = uint8ArrayToString(await command.output()).trim();
    command.close();

    assertEquals(output, VERSION);
  });

  it("can create templates from local folders", async (t) => {
    const target = path.join(
      cwd,
      "tests/fixtures/tmp",
      crypto.randomUUID(),
    );
    needsRemoval.push(target);
    const command = run(["./tests/fixtures/base", target], {
      stdout: "piped",
      stderr: "piped",
    });

    const [_output, _error] = await Promise.all([
      command.output(),
      command.stderrOutput(),
    ]);
    command.close();

    await snapshotDirectory({ t, cwd: target, dot: true, junk: true });
  });

  it("fails when run with too many arguments", async () => {
    const target = path.join(
      cwd,
      "tests/fixtures/tmp",
      crypto.randomUUID(),
    );
    needsRemoval.push(target);
    const command = run(["./tests/fixtures/base", target, "one-too-many"], {
      stdout: "piped",
      stderr: "piped",
    });

    const [_, status, error] = await Promise.all([
      command.output(),
      command.status(),
      command.stderrOutput(),
    ]);
    command.close();

    assertStringIncludes(
      uint8ArrayToString(error),
      "error: Too many arguments: one-too-many",
    );

    assertEquals(status.code, 1);
  });
});
