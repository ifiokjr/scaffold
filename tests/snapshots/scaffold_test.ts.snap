export const snapshot = {};

snapshot[`global > scaffold > can create templates from local folders 1`] = `
{
  ".gitignore": "NOTHING",
  ".should-include": "",
  "asdf/index.js": 'import * as fs from "node:fs/promises";\\n\\nexport function asdf(filepath) {\\n  return fs.writeFile(filepath, "asdf");\\n}\\n',
  "asdf/package.json": '{\\n  "name": "asdf",\\n  "version": "1.0.0",\\n  "type": "module",\\n  "exports": "./index.js"\\n}\\n',
  file: "",
  "hello.text": "hello world",
  "package.json": '{\\n  "name": "scaffold",\\n  "description": "A scaffold for Deno",\\n  "dependencies": {\\n    "asdf": "^1.2.2"\\n  },\\n  "version": "1.0.0"\\n}\\n',
  "rename-thee.md": "Should be renamed from \`rename-me.md\` to \`rename-thee.md\`.\\n",
  "scaffold.json": "{}\\n",
}
`;
